chain/network/src/routing/routing.rs

```diff
@@ -65,7 +65,95 @@ pub enum EdgeType {
     Removed,
 }
 
-pub type Edge = Arc<EdgeInner>;
+#[derive(BorshSerialize, BorshDeserialize, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "test_features", derive(Serialize, Deserialize))]
+pub struct Edge(pub Arc<EdgeInner>);
+
+impl Edge {
+    /// Create an addition edge.
+    pub fn new(
+        peer0: PeerId,
+        peer1: PeerId,
+        nonce: u64,
+        signature0: Signature,
+        signature1: Signature,
+    ) -> Self {
+        Edge(Arc::new(EdgeInner::new(peer0, peer1, nonce, signature0, signature1)))
+    }
+
+    pub fn make_fake_edge(peer0: PeerId, peer1: PeerId, nonce: u64) -> Self {
+        Self(Arc::new(EdgeInner {
+            key: (peer0, peer1),
+            nonce,
+            signature0: Signature::empty(KeyType::ED25519),
+            signature1: Signature::empty(KeyType::ED25519),
+            removal_info: None,
+        }))
+    }
+
+    /// Build a new edge with given information from the other party.
+    pub fn build_with_secret_key(
+        peer0: PeerId,
+        peer1: PeerId,
+        nonce: u64,
+        secret_key: &SecretKey,
+        signature1: Signature,
+    ) -> Self {
+        let hash = if peer0 < peer1 {
+            Self::build_hash(&peer0, &peer1, nonce)
+        } else {
+            Self::build_hash(&peer1, &peer0, nonce)
+        };
+        let signature0 = secret_key.sign(hash.as_ref());
+        Self::new(peer0, peer1, nonce, signature0, signature1)
+    }
+
+    /// Build the hash of the edge given its content.
+    /// It is important that peer0 < peer1 at this point.
+    fn build_hash(peer0: &PeerId, peer1: &PeerId, nonce: u64) -> CryptoHash {
+        let mut buffer = Vec::<u8>::new();
+        let peer0: Vec<u8> = peer0.into();
+        buffer.extend_from_slice(peer0.as_slice());
+        let peer1: Vec<u8> = peer1.into();
+        buffer.extend_from_slice(peer1.as_slice());
+        buffer.write_u64::<LittleEndian>(nonce).unwrap();
+        hash(buffer.as_slice())
+    }
+
+    pub fn make_key(peer0: PeerId, peer1: PeerId) -> (PeerId, PeerId) {
+        if peer0 < peer1 {
+            (peer0, peer1)
+        } else {
+            (peer1, peer0)
+        }
+    }
+
+    /// Helper function when adding a new edge and we receive information from new potential peer
+    /// to verify the signature.
+    pub fn partial_verify(peer0: PeerId, peer1: PeerId, edge_info: &EdgeInfo) -> bool {
+        let pk = peer1.public_key();
+        let (peer0, peer1) = Edge::make_key(peer0, peer1);
+        let data = Edge::build_hash(&peer0, &peer1, edge_info.nonce);
+        edge_info.signature.verify(data.as_ref(), &pk)
+    }
+
+    /// Next nonce of valid addition edge.
+    pub fn next_nonce(nonce: u64) -> u64 {
+        if nonce % 2 == 1 {
+            nonce + 2
+        } else {
+            nonce + 1
+        }
+    }
+}
+
+impl std::ops::Deref for Edge {
+    type Target = EdgeInner;
```

Comment:
It feels stange that both `Edge` and `EdgeInner` are public, and this operates via deref. 

Id probably expect Edge to be public, `EdgeInner` to be private, and `Edge` to have a bunch of accessor functions.

Reply:
Ok, I created another PR for that. https://github.com/near/nearcore/pull/5267